Zend/zend_ast.c

```diff
@@ -702,6 +702,71 @@ ZEND_API zend_result ZEND_FASTCALL zend_ast_evaluate_inner(
 			}
 			zval_ptr_dtor_nogc(&op1);
 			break;
+		case ZEND_AST_CAST:
+			if (UNEXPECTED(zend_ast_evaluate_ex(&op1, ast->child[0], scope, &short_circuited, ctx) != SUCCESS)) {
+				ret = FAILURE;
+				break;
+			}
+			if (ast->attr == Z_TYPE(op1)) {
+				ZVAL_COPY_VALUE(result, &op1);
+			} else {
+				switch (ast->attr) {
+					case _IS_BOOL:
+						ZVAL_BOOL(result, zend_is_true(&op1));
+						zval_ptr_dtor_nogc(&op1);
+						break;
+					case IS_LONG:
+						ZVAL_LONG(result, zval_get_long_func(&op1, false));
+						zval_ptr_dtor_nogc(&op1);
+						break;
+					case IS_DOUBLE:
+						ZVAL_DOUBLE(result, zval_get_double_func(&op1));
+						zval_ptr_dtor_nogc(&op1);
+						break;
+					case IS_STRING:
+						ZVAL_STR(result, zval_get_string_func(&op1));
+						zval_ptr_dtor_nogc(&op1);
+						break;
+					case IS_ARRAY:
+						/* Adapted from VM */
+						if (Z_TYPE(op1) != IS_OBJECT || Z_OBJCE(op1) == zend_ce_closure) {
+							if (Z_TYPE(op1) != IS_NULL) {
+								ZVAL_ARR(result, zend_new_array(1));
+								zend_hash_index_add_new(Z_ARRVAL_P(result), 0, &op1);
+							} else {
+								ZVAL_EMPTY_ARRAY(result);
+							}
+						} else {
+							/* Constant AST only allows building stdClass objects, closures, or user classes;
+							 * all of which should be compatible. */
+							ZEND_ASSERT(ZEND_STD_BUILD_OBJECT_PROPERTIES_ARRAY_COMPATIBLE(&op1));
+
+							/* Optimized version without rebuilding properties HashTable */
+							ZVAL_ARR(result, zend_std_build_object_properties_array(Z_OBJ(op1)));
+							zval_ptr_dtor_nogc(&op1);
+						}
+						break;
+					case IS_OBJECT:
+						/* Adapted from VM */
```

Comment:
I think the object cast is problematic. This will lead to an incorrect result.

```php
class C {
    public $prop = (array) [];
}

$c = new C;
$c->prop->x = 42;
var_dump(new C);
```

Reply:
I remember from the RECV_INIT case that we always re-evaluate refcounted values (https://github.com/php/php-src/blob/ac9392b855cc4db6792a8379223e1fd1bbebd157/Zend/zend_vm_def.h#L5728), of course this particular situation is different but I assumed this held in general.

The code your provided doesn't execute though. I presume you meant `(object) []`, in which case there is indeed a problem.

Reply:
We could ban the object cast from constant evaluation (which should be rare anyway), or find a workaround. I think the former (banning it) seems more reasonable. WDYT?

Reply:
I agree that object cast should simply be forbidden. This feature is not worth changing how we evaluate property constant expressions.

Reply:
@iluuu1994 Should it not be treated the same than https://wiki.php.net/rfc/new_in_initializers simply? At least that's what I'd expect.

Reply:
@bwoebi If @nielsdos wants to add this support, I'm fine with it. Importantly, `zend_ast_evaluate_ctx.had_side_effects` still needs to be set to `false` so that we don't have erroneous caching for `RECV_INIT`.

Reply:
GH actions broke, but I did this now consistent with that RFC (I hope)